### PR TITLE
Add an example on how to add a new device out of tree

### PR DESCRIPTION
**Pull request by albanD**
This is composed of:
- An "has a" torch dispatch class that holds onto the user raw data (a numpy array here for simplicity, but it can be any python/c++ object).
- A torch function mode to properly capture:
  - factory functions and handle the newly added device.
  - cross device functions

Fixes https://github.com/albanD/subclass_zoo/issues/35

### CODE REVIEW

File: new_device.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+aten = torch.ops.aten
+
+# 1. A Tensor that stores custom raw_data and implement functions for it
+class MyDeviceTensor(torch.Tensor):
+    IMPLEMENTATIONS = {}
+
+    @staticmethod
+    def __new__(cls, size, dtype, raw_data=None, requires_grad=False):
+        # Use a meta Tensor here to be used as the wrapper
+        return torch.Tensor._make_subclass(cls, torch.empty(size, dtype=dtype, device="meta"), require_grad=requires_grad)
```
Comment (ezyang) [inline new_device.py:16]: No way to set the device alas!

File: new_device.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    def __init__(self, size, dtype, raw_data=None, requires_grad=False):
+        # Store any provided user raw_data
+        self.raw_data = raw_data
+
+    def __repr__(self):
+        st = super().__repr__()
+        st = st.replace("device='meta'", "device='my_device'")
```
Comment (ezyang) [inline new_device.py:24]: This looks more confusing than hell ha ha ha

File: new_device.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        if func in cls.IMPLEMENTATIONS:
+            try:
+                def super_fn(*args, **kwargs):
+                    return super(cls, MyDeviceTensor).__torch_dispatch__(func, types, args, kwargs)
+                return cls.IMPLEMENTATIONS[func](super_fn, *args, **kwargs or {})
+            except Exception as e:
+                print(e)
+                raise e
+        raise RuntimeError(f"No implementation for 'my_device' for {func}, {args}, {kwargs}")
```
Comment (ezyang) [inline new_device.py:40]: If you use some of the tech from @eellison's fake tensor you should also be able to get the device property to report correctly as well

**Comment by ezyang** [comment]
cc @bdhirsh

**Review by ezyang — APPROVED** [review]

File: new_device.py (hunk 1 above)
Comment (albanD) [reply new_device.py:16]: No, all the C++ code that sets "device" will make sure it is one of the ones in our hardcoded list :(

File: new_device.py (hunk 3 above)
Comment (albanD) [reply new_device.py:40]: I haven't looked into that one yet.
The problem is that even if that tries to create a `torch.device` based on the result from c++, that will fail.

File: new_device.py (hunk 2 above)
Comment (albanD) [reply new_device.py:24]: Yes, this most likely will need to be tweaked by real backends. This is just to be able to easily re-use the plain Tensor print.

**Pull request merged**